clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -370,6 +370,60 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
         }
     }
 
+    public static PartialDefaultRecord readPartiallyFrom(DataInput input,
+                                                         long baseOffset,
+                                                         long baseTimestamp,
+                                                         int baseSequence,
+                                                         Long logAppendTime) throws IOException {
+        int sizeOfBodyInBytes = ByteUtils.readVarint(input);
+        int totalSizeInBytes = ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes;
+        return readPartiallyFrom(input, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
+                baseSequence, logAppendTime);
+    }
+
+    private static PartialDefaultRecord readPartiallyFrom(DataInput input,
+                                                          int sizeInBytes,
+                                                          int sizeOfBodyInBytes,
+                                                          long baseOffset,
+                                                          long baseTimestamp,
+                                                          int baseSequence,
+                                                          Long logAppendTime) throws IOException {
+        try {
+            byte attributes = input.readByte();
+            int skipBytes = 1;
+            long timestampDelta = ByteUtils.readVarlong(input);
+            long timestamp = baseTimestamp + timestampDelta;
+            if (logAppendTime != null)
+                timestamp = logAppendTime;
+            skipBytes += ByteUtils.sizeOfVarlong(timestampDelta);
+
+            int offsetDelta = ByteUtils.readVarint(input);
+            long offset = baseOffset + offsetDelta;
+            int sequence = baseSequence >= 0 ?
+                    DefaultRecordBatch.incrementSequence(baseSequence, offsetDelta) :
+                    RecordBatch.NO_SEQUENCE;
+            skipBytes += ByteUtils.sizeOfVarlong(offsetDelta);
+
+            int keySize = ByteUtils.readVarint(input);
+            boolean hasKey = false;
+            if (keySize >= 0) {
+                hasKey = true;
+            }
+            skipBytes += ByteUtils.sizeOfVarlong(keySize);
+
+            skipBytes = sizeOfBodyInBytes - skipBytes;
```

Comment:
Seems we are skipping validation of the value size and header. That seems dangerous. Seems we could do the validation and still keep the benefit of skipping allocations.

Reply:
Hmm. For value size, we do not check if it is always larger than -1: i.e. even if it is less than -1 we do not do anything.

For header size, yeah we could probably do it as well.

Reply:
@hachikuji I looked into how skipBytes in compressed input is implemented, and it looks like this:

```
long remaining = n;
        int nr;

        if (n <= 0) {
            return 0;
        }

        int size = (int)Math.min(MAX_SKIP_BUFFER_SIZE, remaining);
        byte[] skipBuffer = new byte[size];
        while (remaining > 0) {
            nr = read(skipBuffer, 0, (int)Math.min(size, remaining));
            if (nr < 0) {
                break;
            }
            remaining -= nr;
        }

        return n - remaining;
```

So it means if your read bytes is not that large (than MAX_SKIP_BUFFER_SIZE = 2048), then skipping N bytes in one call would be cheaper than skipping twice with N/2 bytes each call. 

Given this, I'm a bit on the fence whether it worth paying the cost just to validate num.headers is valid. WDYT?

Reply:
@guozhangwang I think we shouldn't relax schema validation. We would have to think through all the implications of accepting bad data. Is the performance really that much worse? I thought the biggest benefit was avoiding the allocations, which we should still be able to do.

Reply:
As we discussed offline, with `MAX_SKIP_BUFFER_SIZE == 2048` it means if the bytes to skip is smaller than 2048 then there would be not much benefits left, and even more, since we'd need to skipBytes three times (key, value, headers) it may be more GC cost compared with allocating a single large byte array for the whole record.